pkg/ddl/partition.go

```diff
@@ -3190,54 +3193,116 @@ func (w *reorgPartitionWorker) BackfillData(handleRange reorgBackfillTask) (task
 		}
 		txn.SetOption(kv.ResourceGroupName, w.jobContext.resourceGroupName)
 
-		rowRecords, nextKey, taskDone, err := w.fetchRowColVals(txn, handleRange)
+		nextKey, taskDone, err := w.fetchRowColVals(txn, handleRange)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		taskCtx.nextKey = nextKey
 		taskCtx.done = taskDone
 
-		warningsMap := make(map[errors.ErrorID]*terror.Error)
-		warningsCountMap := make(map[errors.ErrorID]int64)
-		for _, prr := range rowRecords {
-			taskCtx.scanCount++
-
-			err = txn.Set(prr.key, prr.vals)
+		var found map[string][]byte
+		// If non-clustered table, then we need to replace the _tidb_rowid handles since
+		// there may be duplicates across different partitions, due to EXCHANGE PARTITION.
+		// Meaning we need to check here if a record was double written to the new partition,
+		// i.e. concurrently written by StateWriteOnly or StateWriteReorganization.
+		// If so, then we must skip it.
+		if len(w.oldKeys) > 0 {
+			// We need to use pessimistic, since we have no other unique key than the _tidb_rowid,
+			// which we are changing, so we need to check if we can write the old one without collision.
+			//txn.SetOption(kv.Pessimistic, true)
+
+			// we must check if old IDs already been written,
+			// i.e. double written by StateWriteOnly or StateWriteReorganization.
+			// The good thing is that we can then also skip the index generation for that row and we don't need to
+			// check if duplicate index entries was already copied either!
+			// If we skip checking, then we will overwrite those double written rows.
+			// TODO: Would it be OK to overwrite them?
+			// They will always be double writing during this state, but can change between fetchRowColVals
+			// and writing here, which should be caught by the SetAssertNotExists below.
+			found, err = txn.BatchGet(ctx, w.oldKeys)
 			if err != nil {
 				return errors.Trace(err)
 			}
-			taskCtx.addedCount++
-			if prr.warning != nil {
-				if _, ok := warningsCountMap[prr.warning.ID()]; ok {
-					warningsCountMap[prr.warning.ID()]++
-				} else {
-					warningsCountMap[prr.warning.ID()] = 1
-					warningsMap[prr.warning.ID()] = prr.warning
-				}
-			}
```

Comment:
warnings were never set, and cannot happen, since we are copying data without any conversions.